File: frontend/src/api/scil.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { api } from './client'

export interface ScilRouteCount {
  route: string
  count: number
  llm_calls: number
}

export interface ScilMetricsSummary {
  total_requests: number
  llm_calls: number
  llm_calls_avoided: number
  cache_hit_rate: number
  retried_turns: number
  retry_success_rate: number
  avg_latency_ms_by_route: Record<string, number>
  routes: ScilRouteCount[]
  hallucination_flags: number
}

export interface ScilTimeseriesPoint {
  date: string
  route: string
  count: number
  llm_calls: number
}

export interface ScilCacheEntry {
  id: number
  agent_id: string
  agent_name: string | null
  input_text: string
  output_type: string
  hit_count: number
  validated: boolean
  created_at: string
  last_hit_at: string | null
  similarity_threshold: number
}

export interface ScilCacheListResponse {
  items: ScilCacheEntry[]
  total: number
  limit: number
  offset: number
}

export interface ScilCacheSimilarityCheckResult {
  similarity: number
  threshold: number
  would_hit: boolean
}

export function useCheckCacheSimilarity() {
  return useMutation({
    mutationFn: (payload: { agent_id: string; text_a: string; text_b: string }) =>
      api.post<ScilCacheSimilarityCheckResult>('/scil/cache/similarity-check', payload),
  })
}

export interface ScilCorrectionEntry {
  id: number
  agent_id: string
  agent_name: string | null
  input_text: string
  error_signature: string
  error_detail: string
  correction_source: string
  reuse_count: number
  created_at: string
}

export interface ScilCorrectionListResponse {
  items: ScilCorrectionEntry[]
  total: number
  limit: number
  offset: number
}

const KEY = ['scil'] as const

export function useScilSummary(rangeDays: number) {
  const from = new Date(Date.now() - rangeDays * 86_400_000).toISOString()
  return useQuery({
    queryKey: [...KEY, 'summary', rangeDays],
    queryFn: () => api.get<ScilMetricsSummary>(`/scil/metrics/summary?from_date=${encodeURIComponent(from)}`),
    refetchInterval: 30_000,
  })
}

export function useScilTimeseries(rangeDays: number) {
  return useQuery({
    queryKey: [...KEY, 'timeseries', rangeDays],
    queryFn: () => api.get<ScilTimeseriesPoint[]>(`/scil/metrics/timeseries?range_days=${rangeDays}`),
    refetchInterval: 30_000,
  })
}

export function useScilCacheEntries(offset: number, limit = 25) {
  return useQuery({
    queryKey: [...KEY, 'cache', offset, limit],
    queryFn: () => api.get<ScilCacheListResponse>(`/scil/cache/entries?limit=${limit}&offset=${offset}`),
    refetchInterval: 30_000,
  })
}

export function useScilCorrections(offset: number, limit = 25, errorSignature?: string) {
  return useQuery({
    queryKey: [...KEY, 'corrections', offset, limit, errorSignature ?? null],
    queryFn: () => {
      const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
      if (errorSignature) params.set('error_signature', errorSignature)
      return api.get<ScilCorrectionListResponse>(`/scil/corrections?${params.toString()}`)
    },
    refetchInterval: 30_000,
  })
}

export function useDeleteCacheEntry() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: number) => api.delete<void>(`/scil/cache/entries/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: KEY }),
  })
}

export function usePurgeAgentCache() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (agentId: string) => api.post<void>('/scil/cache/purge', { agent_id: agentId }),
    onSuccess: () => qc.invalidateQueries({ queryKey: KEY }),
  })
}

export function useDeleteCorrection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: number) => api.delete<void>(`/scil/corrections/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: KEY }),
  })
}

// --- Eval framework: golden-question regression suite ----------------------

export interface ScilEvalCaseEntry {
  id: number
  agent_id: string
  question: string
  expected_criteria: string
  is_active: boolean
  created_at: string
  last_passed: boolean | null
  last_run_at: string | null
}

export interface ScilEvalRunResult {
  case_id: number
  question: string
  passed: boolean
  actual_response: string
  judge_reasoning: string
  latency_ms: number
}

export interface ScilEvalBatchSummary {
  batch_id: string
  agent_id: string
  total: number
  passed: number
  results: ScilEvalRunResult[]
  created_at: string
}

export function useEvalCases(agentId: string | undefined) {
  return useQuery({
    queryKey: [...KEY, 'eval-cases', agentId],
    queryFn: () => api.get<ScilEvalCaseEntry[]>(`/scil/eval/cases?agent_id=${agentId}`),
    enabled: !!agentId,
  })
}

export function useCreateEvalCase() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (payload: { agent_id: string; question: string; expected_criteria: string }) =>
      api.post<ScilEvalCaseEntry>('/scil/eval/cases', payload),
    onSuccess: () => qc.invalidateQueries({ queryKey: [...KEY, 'eval-cases'] }),
  })
}

export function useDeleteEvalCase() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: number) => api.delete<void>(`/scil/eval/cases/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: [...KEY, 'eval-cases'] }),
  })
}

export function useLatestEvalBatch(agentId: string | undefined) {
  return useQuery({
    queryKey: [...KEY, 'eval-runs-latest', agentId],
    queryFn: () => api.get<ScilEvalBatchSummary>(`/scil/eval/runs/latest?agent_id=${agentId}`),
    enabled: !!agentId,
    retry: false,
  })
}

export function useRunEvalBatch() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (agentId: string) => api.post<ScilEvalBatchSummary>('/scil/eval/run', { agent_id: agentId }),
    onSuccess: (_data, agentId) => {
      qc.invalidateQueries({ queryKey: [...KEY, 'eval-cases', agentId] })
      qc.invalidateQueries({ queryKey: [...KEY, 'eval-runs-latest', agentId] })
    },
  })
}

// --- Eval framework: sampled live-traffic groundedness ----------------------

export interface ScilGroundednessSummaryPoint {
  date: string
  grounded: number
  ungrounded: number
}

export interface ScilGroundednessSampleEntry {
  id: number
  agent_id: string
  agent_name: string | null
  input_text: string
  grounded: boolean
  reason: string | null
  created_at: string
}

export interface ScilGroundednessSummary {
  total_samples: number
  grounded_rate: number
  timeseries: ScilGroundednessSummaryPoint[]
  recent_flagged: ScilGroundednessSampleEntry[]
}

export function useGroundednessSummary(rangeDays: number, agentId?: string) {
  return useQuery({
    queryKey: [...KEY, 'groundedness-summary', rangeDays, agentId ?? null],
    queryFn: () => {
      const params = new URLSearchParams({ range_days: String(rangeDays) })
      if (agentId) params.set('agent_id', agentId)
      return api.get<ScilGroundednessSummary>(`/scil/eval/groundedness/summary?${params.toString()}`)
    },
    refetchInterval: 30_000,
  })
}

File: backend/app/schemas/scil.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class ScilCacheEntry(BaseModel):
    id: int
    agent_id: uuid.UUID
    agent_name: str | None
    input_text: str
    output_type: str
    hit_count: int
    validated: bool
    created_at: datetime
    last_hit_at: datetime | None
    # This entry's owning agent's CURRENT model_config.scil.cache_similarity_
    # threshold (not a value frozen at write time — an agent's threshold can
    # change after this row was written, and the dashboard should always show
    # what a NEW question would be compared against, not history).
    similarity_threshold: float


class ScilCacheSimilarityCheckRequest(BaseModel):
    agent_id: uuid.UUID
    text_a: str
    text_b: str


class ScilCacheSimilarityCheckResponse(BaseModel):
    similarity: float
    threshold: float
    would_hit: bool


class ScilCacheListResponse(BaseModel):
    items: list[ScilCacheEntry]
    total: int
    limit: int
    offset: int


class ScilRouteCount(BaseModel):
    route: str
    count: int
    llm_calls: int


class ScilMetricsSummary(BaseModel):
    total_requests: int
    llm_calls: int
    llm_calls_avoided: int
    cache_hit_rate: float
    # Self-correction loop (route="llm_retry"): how many turns needed at
    # least one validation retry, and what fraction of those ended in a
    # validated answer (measured against auto_retry correction-memory
    # writes, which only happen on recovery).
    retried_turns: int
    retry_success_rate: float
    avg_latency_ms_by_route: dict[str, float]
    routes: list[ScilRouteCount]
    # Count of scil_correction_memory rows in this range whose error_signature
    # starts with "Hallucination:" — corrected hallucinations only. Turns that
    # exhausted their retries still hallucinating are NOT counted here (they
    # never got a correction-memory row); those are visible as
    # hallucination_unresolved events in the Debug Console/RCA instead.
    hallucination_flags: int


class ScilPurgeRequest(BaseModel):
    agent_id: uuid.UUID


class ScilCorrectionEntry(BaseModel):
    id: int
    agent_id: uuid.UUID
    agent_name: str | None
    input_text: str
    error_signature: str
    error_detail: str
    correction_source: str
    reuse_count: int
    created_at: datetime


class ScilCorrectionListResponse(BaseModel):
    items: list[ScilCorrectionEntry]
    total: int
    limit: int
    offset: int


class ScilEntityEntry(BaseModel):
    id: int
    agent_id: uuid.UUID
    agent_name: str | None
    entity_text: str
    use_count: int
    created_at: datetime
    last_used_at: datetime | None


class ScilEntityListResponse(BaseModel):
    items: list[ScilEntityEntry]
    total: int
    limit: int
    offset: int


class ScilTimeseriesPoint(BaseModel):
    """One (day, route) bucket — the dashboard pivots these into per-route
    series for the savings-over-time chart."""

    date: str
    route: str
    count: int
    llm_calls: int


class ScilEvalRunRequest(BaseModel):
    agent_id: uuid.UUID


class ScilEvalCaseCreate(BaseModel):
    agent_id: uuid.UUID
    question: str
    expected_criteria: str


class ScilEvalCaseEntry(BaseModel):
    id: int
    agent_id: uuid.UUID
    question: str
    expected_criteria: str
    is_active: bool
    created_at: datetime
    # Denormalized from this case's most recent scil_eval_runs row, if any
    # batch has run yet — lets the case list show current pass/fail without
    # a second round trip per case.
    last_passed: bool | None = None
    last_run_at: datetime | None = None


class ScilEvalRunResult(BaseModel):
    case_id: int
    question: str
    passed: bool
    actual_response: str
    judge_reasoning: str
    latency_ms: int


class ScilEvalBatchSummary(BaseModel):
    batch_id: uuid.UUID
    agent_id: uuid.UUID
    total: int
    passed: int
    results: list[ScilEvalRunResult]
    created_at: datetime


class ScilGroundednessSummaryPoint(BaseModel):
    date: str
    grounded: int
    ungrounded: int


class ScilGroundednessSampleEntry(BaseModel):
    id: int
    agent_id: uuid.UUID
    agent_name: str | None
    input_text: str
    grounded: bool
    reason: str | None
    created_at: datetime


class ScilGroundednessSummary(BaseModel):
    total_samples: int
    grounded_rate: float
    timeseries: list[ScilGroundednessSummaryPoint]
    recent_flagged: list[ScilGroundednessSampleEntry]
